File: sdk/go/pkg/dgraph/dgraph.go
/*
 * Copyright 2024 Hypermode Inc.
 * Licensed under the terms of the Apache License, Version 2.0
 * See the LICENSE file that accompanied this code for further details.
 *
 * SPDX-FileCopyrightText: 2024 Hypermode Inc. <hello@hypermode.com>
 * SPDX-License-Identifier: Apache-2.0
 */

package dgraph

import (
	"encoding/json"
	"errors"
	"fmt"
	"strings"
)

// Represents a Dgraph request.
type Request struct {
	Query     *Query
	Mutations []*Mutation
}

// Represents a Dgraph query.
type Query struct {
	Query     string
	Variables map[string]string
}

// Represents a Dgraph mutation.
type Mutation struct {

	// JSON for setting data
	SetJson string

	// JSON for deleting data
	DelJson string

	// RDF N-Quads for setting data
	SetNquads string

	// RDF N-Quads for deleting data
	DelNquads string

	// Condition for the mutation, as a DQL @if directive
	Condition string
}

// Represents a Dgraph response.
type Response struct {
	Json string
	Uids map[string]string
}

// Creates a new Dgraph query.
func NewQuery(query string) *Query {
	return &Query{
		Query:     query,
		Variables: make(map[string]string),
	}
}

// Adds a variable to the query.
func (q *Query) WithVariable(key string, value any) *Query {
	switch value := value.(type) {
	case string:
		// strings are passed as-is, without extra encoding
		q.Variables[key] = value
	case int, int8, int16, int32, int64,
		uint, uint8, uint16, uint32, uint64,
		float32, float64, bool,
		[]float32:
		s, _ := json.Marshal(value)
		q.Variables[key] = string(s)
	default:
		panic(fmt.Errorf("unsupported DQL variable type: %T (must be string, integer, float, bool, or []float32)", value))
	}

	return q
}

// Creates a new Dgraph mutation.
func NewMutation() *Mutation {
	return &Mutation{}
}

// Adds a JSON string for setting data to the mutation.
func (m *Mutation) WithSetJson(json string) *Mutation {
	m.SetJson = json
	return m
}

// Adds a JSON string for deleting data to the mutation.
func (m *Mutation) WithDelJson(json string) *Mutation {
	m.DelJson = json
	return m
}

// Adds RDF N-Quads for setting data to the mutation.
func (m *Mutation) WithSetNquads(nquads string) *Mutation {
	m.SetNquads = nquads
	return m
}

// Adds RDF N-Quads for deleting data to the mutation.
func (m *Mutation) WithDelNquads(nquads string) *Mutation {
	m.DelNquads = nquads
	return m
}

// Adds a condition to the mutation, as a DQL @if directive.
func (m *Mutation) WithCondition(cond string) *Mutation {
	m.Condition = cond
	return m
}

// Deprecated: Use ExecuteQuery or ExecuteMutations instead.
func Execute(connection string, request *Request) (*Response, error) {
	response := hostExecuteQuery(&connection, request)
	if response == nil {
		return nil, errors.New("Failed to execute the DQL query.")
	}

	return response, nil
}

// Executes a DQL query on the Dgraph database, optionally with mutations.
func ExecuteQuery(connection string, query *Query, mutations ...*Mutation) (*Response, error) {
	request := &Request{
		Query:     query,
		Mutations: mutations,
	}

	return Execute(connection, request)
}

// Executes one or more mutations on the Dgraph database.
func ExecuteMutations(connection string, mutations ...*Mutation) (*Response, error) {
	request := &Request{
		Mutations: mutations,
	}

	return Execute(connection, request)
}

// Alters the schema of the dgraph database
func AlterSchema(connection, schema string) error {
	resp := hostAlterSchema(&connection, &schema)
	if resp == nil {
		return errors.New("Failed to alter the schema.")
	}

	return nil
}

// Drops an attribute from the schema.
func DropAttr(connection, attr string) error {
	response := hostDropAttribute(&connection, &attr)
	if response == nil {
		return errors.New("Failed to drop the attribute.")
	}

	return nil
}

// Drops all data from the database.
func DropAll(connection string) error {
	response := hostDropAllData(&connection)
	if response == nil {
		return errors.New("Failed to drop all data.")
	}

	return nil
}

// Ensures proper escaping of RDF string literals
func EscapeRDF(value string) string {
	var sb strings.Builder
	for _, r := range value {
		switch r {
		case '\\':
			sb.WriteString(`\\`)
		case '"':
			sb.WriteString(`\"`)
		case '\n':
			sb.WriteString(`\n`)
		case '\r':
			sb.WriteString(`\r`)
		case '\t':
			sb.WriteString(`\t`)
		default:
			// handle control characters
			if r < 0x20 || (r >= 0x7F && r <= 0x9F) {
				sb.WriteString(fmt.Sprintf(`\u%04X`, r))
			} else {
				sb.WriteRune(r)
			}
		}
	}
	return sb.String()
}

File: sdk/assemblyscript/src/assembly/dgraph.ts
/*
 * Copyright 2024 Hypermode Inc.
 * Licensed under the terms of the Apache License, Version 2.0
 * See the LICENSE file that accompanied this code for further details.
 *
 * SPDX-FileCopyrightText: 2024 Hypermode Inc. <hello@hypermode.com>
 * SPDX-License-Identifier: Apache-2.0
 */

import { JSON } from "json-as";
import * as utils from "./utils";

// @ts-expect-error: decorator
@external("modus_dgraph_client", "executeQuery")
declare function hostExecuteQuery(
  connection: string,
  request: Request,
): Response;

// @ts-expect-error: decorator
@external("modus_dgraph_client", "alterSchema")
declare function hostAlterSchema(connection: string, schema: string): string;

// @ts-expect-error: decorator
@external("modus_dgraph_client", "dropAttribute")
declare function hostDropAttribute(connection: string, attr: string): string;

// @ts-expect-error: decorator
@external("modus_dgraph_client", "dropAllData")
declare function hostDropAllData(connection: string): string;

/**
 * @deprecated Use `executeQuery` or `executeMutations` instead,
 * which don't require creating a `Request` object explicitly.
 */
export function execute(connection: string, request: Request): Response {
  return _execute(connection, request);
}

/**
 * Executes a DQL query on the Dgraph database, optionally with mutations.
 * @param connection - the name of the connection
 * @param query - the query to execute
 * @param mutations - the mutations to execute, if any
 * @returns The response from the Dgraph server
 * @remarks This function is a convenience wrapper around the `execute` function.
 */
export function executeQuery(
  connection: string,
  query: Query,
  ...mutations: Mutation[]
): Response {
  const request = new Request(query, mutations);
  return _execute(connection, request);
}

/**
 * Executes one or more mutations on the Dgraph database.
 * @param connection - the name of the connection
 * @param mutations - the mutations to execute
 * @returns The response from the Dgraph server
 * @remarks This function is a convenience wrapper around the `execute` function.
 */
export function executeMutations(
  connection: string,
  ...mutations: Mutation[]
): Response {
  const request = new Request(null, mutations);
  return _execute(connection, request);
}

// @ts-expect-error: decorator
@inline
function _execute(connection: string, request: Request): Response {
  const response = hostExecuteQuery(connection, request);
  if (!response) {
    throw new Error("Error executing DQL.");
  }

  return response;
}

/**
 * Alters the schema of the dgraph database
 * @param connection - the name of the connection
 * @param schema - the schema to alter
 * @returns The response from the Dgraph server
 */
export function alterSchema(connection: string, schema: string): string {
  const response = hostAlterSchema(connection, schema);
  if (utils.resultIsInvalid(response)) {
    throw new Error("Error invoking DQL.");
  }

  return response;
}

/**
 * Drops an attribute from the schema.
 * @param connection - the name of the connection
 * @param attr - the attribute to drop
 * @returns The response from the Dgraph server
 */
export function dropAttr(connection: string, attr: string): string {
  const response = hostDropAttribute(connection, attr);
  if (utils.resultIsInvalid(response)) {
    throw new Error("Error invoking DQL.");
  }

  return response;
}

/**
 * Drops all data from the database.
 * @param connection - the name of the connection
 * @returns The response from the Dgraph server
 */
export function dropAll(connection: string): string {
  const response = hostDropAllData(connection);
  if (utils.resultIsInvalid(response)) {
    throw new Error("Error invoking DQL.");
  }

  return response;
}

/**
 * Represents a Dgraph request.
 */
export class Request {
  constructor(Query: Query | null = null, Mutations: Mutation[] | null = null) {
    if (Query) {
      this.query = Query;
    }
    if (Mutations) {
      this.mutations = Mutations;
    }
  }
  query: Query = new Query();
  mutations: Mutation[] = [];
}

/**
 * Represents a Dgraph query.
 */
export class Query {
  constructor(query: string = "", variables: Variables = new Variables()) {
    this.query = query;
    this.variables = variables.toMap();
  }
  query: string = "";
  variables: Map<string, string> = new Map<string, string>();

  /**
   * Adds a variable to the query.
   */
  withVariable<T>(name: string, value: T): this {
    if (isString<T>()) {
      this.variables.set(name, value as string);
      return this;
    } else if (isInteger<T>()) {
      this.variables.set(name, JSON.stringify(value));
      return this;
    } else if (isFloat<T>()) {
      this.variables.set(name, JSON.stringify(value));
      return this;
    } else if (isBoolean<T>()) {
      this.variables.set(name, JSON.stringify(value));
      return this;
    } else {
      throw new Error(
        "Unsupported DQL variable type. Must be string, integer, float, or boolean.",
      );
    }
  }
}

/**
 * Represents a Dgraph mutation.
 */
export class Mutation {
  /**
   * Creates a new Mutation object.
   * @param setJson - JSON for setting data
   * @param delJson - JSON for deleting data
   * @param setNquads - RDF N-Quads for setting data
   * @param delNquads - RDF N-Quads for deleting data
   * @param condition - Condition for the mutation, as a DQL `@if` directive
   *
   * @remarks - All parameters are optional.
   * For clarity, prefer passing no arguments here, but use the `with*` methods instead.
   */
  constructor(
    public setJson: string = "",
    public delJson: string = "",
    public setNquads: string = "",
    public delNquads: string = "",
    public condition: string = "",
  ) {}

  /**
   * Adds a JSON string for setting data to the mutation.
   */
  withSetJson(json: string): Mutation {
    this.setJson = json;
    return this;
  }

  /**
   * Adds a JSON string for deleting data to the mutation.
   */
  withDelJson(json: string): Mutation {
    this.delJson = json;
    return this;
  }

  /**
   * Adds RDF N-Quads for setting data to the mutation.
   */
  withSetNquads(nquads: string): Mutation {
    this.setNquads = nquads;
    return this;
  }

  /**
   * Adds RDF N-Quads for deleting data to the mutation.
   */
  withDelNquads(nquads: string): Mutation {
    this.delNquads = nquads;
    return this;
  }

  /**
   * Adds a condition to the mutation, as a DQL `@if` directive.
   */
  withCondition(cond: string): Mutation {
    this.condition = cond;
    return this;
  }
}

/**
 * Represents a Dgraph response.
 */
export class Response {
  Json: string = "";
  Uids: Map<string, string> | null = null;
}

/**
 * Represents a set of variables to be used in a Dgraph query.
 */
export class Variables {
  private data: Map<string, string> = new Map<string, string>();

  public set<T>(name: string, value: T): void {
    if (isString<T>()) {
      this.data.set(name, value as string);
      return;
    } else if (isInteger<T>()) {
      this.data.set(name, JSON.stringify(value));
      return;
    } else if (isFloat<T>()) {
      this.data.set(name, JSON.stringify(value));
      return;
    } else if (isBoolean<T>()) {
      this.data.set(name, JSON.stringify(value));
      return;
    } else {
      throw new Error(
        "Unsupported DQL variable type. Must be string, integer, float, or boolean.",
      );
    }
  }

  public toMap(): Map<string, string> {
    return this.data;
  }
}

/**
 * Ensures proper escaping of RDF string literals
 */
export function escapeRDF(value: string): string {
  let output = "";
  for (let i = 0; i < value.length; i++) {
    const cc = value.charCodeAt(i);
    switch (cc) {
      case 0x5c:
        output += "\\\\";
        break;
      case 0x22:
        output += '\\"';
        break;
      case 0x0a:
        output += "\\n";
        break;
      case 0x0d:
        output += "\\r";
        break;
      case 0x09:
        output += "\\t";
        break;
      default: {
        // handle control characters
        if (cc < 0x20 || (cc >= 0x7f && cc < 0x9f)) {
          output += "\\u" + cc.toString(16).padStart(4, "0");
        } else {
          output += value.charAt(i);
        }
      }
    }
  }

  return output;
}
